pkg/daemon/server/service/pipeline_metrics_query.go

```diff
@@ -233,6 +239,57 @@ func (ps *pipelineMetadataQuery) GetVertexMetrics(ctx context.Context, req *daem
 	return resp, nil
 }
 
+func (ps *pipelineMetadataQuery) GetPipelineStatus(ctx context.Context, req *daemon.GetPipelineStatusRequest) (*daemon.GetPipelineStatusResponse, error) {
+
+	resp := new(daemon.GetPipelineStatusResponse)
+
+	// get all vertices of pipeline
+	vertices := ps.pipeline.Spec.Vertices
+
+	// loop over vertices and get metrics to check pending messages vs processing rate
+	for _, vertex := range vertices {
+		vertexReq := new(daemon.GetVertexMetricsRequest)
+		vertexReq.Vertex = &vertex.Name
+		vertexResp, err := ps.GetVertexMetrics(ctx, vertexReq)
+		// if err is not nil, more than likely autoscaling is down to 0 and metrics are not available
+		if err != nil {
+			resp.Status = &daemon.PipelineStatus{
+				Status:  pointer.String(PipelineStatusUnknown),
+				Message: pointer.String("Pipeline status is unknown."),
+			}
+			return resp, nil
+		}
+
+		// may need to revisit later, another concern could be that the processing rate is too slow instead of just 0
+		for _, vertexMetrics := range vertexResp.VertexMetrics {
+			if pending, ok := vertexMetrics.GetPendings()["default"]; ok {
+				if processingRate, ok := vertexMetrics.GetProcessingRates()["default"]; ok {
+					if pending > 0 && processingRate == 0 {
+						resp.Status = &daemon.PipelineStatus{
+							Status:  pointer.String(PipelineStatusError),
+							Message: pointer.String(fmt.Sprintf("Pipeline has an error. Vertex %s is not processing pending messages.", vertex.Name)),
+						}
+						return resp, nil
+					}
+				}
+			} else {
```

Comment:
We don't need this `else`.

It is expected sometimes there's no `pending` information:

1. Source like `HTTP`, it does not have `pending` at all;
2. Based on the `GetVertexMetrics` implementation, when scaling down to 0 replica, the `pending` is also not available.

For these cases, we should consider it's ok.